### PR TITLE
Self Assign Roles

### LINKED ISSUE #136: Role-related functionality
### Feature Description:
Make a substitute for .inrole, etc. to reduce nadeko dependency. I can probably do it myself when I get around to it.

### Implementation Suggestion:
I'll read the `discord.py` docs later.

### PR DESCRIPTION
**Pull request by krubenok**
## Description

I have created basic roles functionality to allow users to self-assign roles to themselves in certain categories. I have also relocated other role related functionality in the project to `roles.py` as it seemed better located there with other similar functions. In addition to the `?iam` generic role command, some commands have been created with limitations such as year and faculty (where a user may have only one role in that category at a time). 

## Motivation and Context

A lot of mod labour is wasted on assigning roles of little importance that are mostly just for community convenience. Letting them be self assigned within reason should save a decent amount of work. 

Related to #136 

## How Has This Been Tested?

Tested in the testing server. Would appreciate more tests with more roles to check for more corner cases. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by krubenok** [comment]
> 3\. There is no reason that the list of assignable roles couldn't be in the database, but I do suppose that that would take more effort and would be of relatively minimal benefit (essentially, it would be beneficial only in letting mods change roles and letting them do so without having to restart Canary or bother the bot's sysop).

This is the intention, but I've punted it to the role creation functionality for the time being it's gonna stick to the config file.

File: config/config.ini [hunk 1]
```diff
@@ -79,3 +79,10 @@ TransactionTax = 0.15
 [Betting]
 RollCases = 66, 90, 99, 100
 RollReturns = 0, 2, 4, 10
+
+[Roles]
+Pronouns = She/Her, He/Him, They/Them
+Fields = Anatomy and Cell Biology, Anthropology, Agriculture, Architecture, Biochemistry, Bioengineering, Biology, Bioresource Engineering, Chemical Engineering, Chemistry, Civil Engineering, Classics, cogito, Commerce, Computer Engineering, Computer Science, Computer Science/Biology, Cultural Studies, Desautels, Economics , Electrical Engineering, English, Experimental Medicine, Finance, Geography, History, Indigenous Studies, International Development Studies, Jewish Studies, mac kid, Materials Engineering, Math, MBA, Mechanical Engineering, Medicine, Microbiology and Immunology, Neuroscience, Nursing, Pharmacology, Philosophy, Physical Therapy, Physics, Physiology, Political Science, Psychiatry, Psychology, Public Health, Social Work, Sociology, Software Engineering, Statistics
+Faculties = Science, Engineering, Management, Arts, ArtSci, Agriculture and Environment, Law, Education, Dentistry, Music
+Years = U0, U1, U2, U3, U4, PhD, Masters, Post Doc, wenthere
+Generics = weeb, weeb stomper
```
Comment (krubenok) [inline config/config.ini:88]: This is just the default suggested set but the set used on our server can obviously differ from this.

File: cogs/roles.py [hunk 2]
```diff
@@ -32,132 +32,89 @@ def __init__(self, bot):
         self.bot = bot
         self.roles = self.bot.config.roles
 
-    @commands.command()
-    async def pronoun(self, ctx, transaction, pronoun):
+    async def toggleRole(self, ctx, transaction, requestedRole, category):
         """
-        Self-assign a pronoun role to a user. 
+        Assigs a single role to a user with no checks from a category of roles
```
Comment (ValerianClerc) [inline cogs/roles.py:37]: Typo: Assigns*

File: cogs/roles.py [hunk 3]
```diff
@@ -32,132 +32,89 @@ def __init__(self, bot):
...
+        if (transaction == "add") and (requestedRole in roles):
+            role = discord.utils.get(ctx.guild.roles, name=requestedRole)
             await member.add_roles(role, reason="Self Requested")
-        elif (transaction == "remove") and (pronoun in pronouns):
-            role = discord.utils.get(ctx.guild.roles, name=pronoun)
+        elif (transaction == "remove") and (requestedRole in roles):
+            role = discord.utils.get(ctx.guild.roles, name=requestedRole)
             await member.remove_roles(role, reason="Self Requested")
-
         else:
```
Comment (ValerianClerc) [inline cogs/roles.py:51]: Should we have an `elif` for a different error message sent back if the transaction is invalid (if it's not "add" or "remove")?

File: cogs/roles.py [hunk 4]
```diff
@@ -32,132 +32,89 @@ def __init__(self, bot):
...
-            await ctx.send("Error, that role doesn't exist")
+            await ctx.send("Error, that role doesn't exist in this category")
+        return
 
-    @commands.command()
-    async def field(self, ctx, transaction, field):
+    async def toggleRestrictedRole(self, ctx, transaction, requestedRole,
```
Comment (ValerianClerc) [inline cogs/roles.py:55]: I kind of like `toggleUniqueRole` a bit better because it seems clearer to me what the purpose of the function is. Up to you though

File: cogs/roles.py (hunk 3 above)
Comment (krubenok) [reply cogs/roles.py:51]: Do I need a different error for add/remove being wrong vs. the role not existing or is one catch all error sufficient?

File: cogs/roles.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                if old_role is not None:
+                    await member.remove_roles(old_role,
+                                              reason="Self Requested")
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove") and (requestedRole in roles):
+            role = discord.utils.get(ctx.guild.roles, name=requestedRole)
+            await member.remove_roles(role, reason="Self Requested")
+        else:
+            await ctx.send("Error, that role doesn't exist in this category")
+        return
```
Comment (davidlougheed) [inline cogs/roles.py:81]: dont need this return

File: cogs/roles.py (hunk 3 above)
Comment (ValerianClerc) [reply cogs/roles.py:51]: I was thinking that it could be nice for the user of the command to know if they typed in a non-existant role or if they just used the command wrong. But yeah this is super subjective

File: cogs/roles.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                old_role = discord.utils.get(ctx.guild.roles,
+                                             name=category_roles)
+                if old_role is not None:
+                    await member.remove_roles(old_role,
+                                              reason="Self Requested")
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove") and (requestedRole in roles):
+            role = discord.utils.get(ctx.guild.roles, name=requestedRole)
+            await member.remove_roles(role, reason="Self Requested")
+        else:
```
Comment (davidlougheed) [inline cogs/roles.py:79]: maybe this should early-return instead of being a final else
`if requestedRole not in roles`
different one for `transaction not in ("add", "remove")` or whatever for a different err msg like what val's talking abt

then we can move 
```python
role = discord.utils.get(ctx.guild.roles, name=requestedRole)
```
out of the blocks so that line isnt duplicated twice

File: cogs/roles.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    def __init__(self, bot):
+        self.bot = bot
+        self.roles = self.bot.config.roles
+
+    async def toggleRole(self, ctx, transaction, requestedRole, category):
+        """
+        Assigs a single role to a user with no checks from a category of roles
+        """
+        member = ctx.message.author
+        roles = self.roles.get(category).split(", ")
```
Comment (davidlougheed) [inline cogs/roles.py:40]: here u can just index with `[]`
or, if we want to prevent key errors, we should do a default that is `""` instead of `None`
since if the get gives None as a default itll just raise exception on split anyway

File: config/config.ini [hunk 8]
```diff
@@ -79,3 +79,10 @@ TransactionTax = 0.15
 [Betting]
 RollCases = 66, 90, 99, 100
 RollReturns = 0, 2, 4, 10
+
+[Roles]
+Pronouns = She/Her, He/Him, They/Them
```
Comment (davidlougheed) [inline config/config.ini:84]: gonna hav eto clear all of these categories with mods sry

File: cogs/roles.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            await ctx.send("Error, that role doesn't exist in this category")
+        return
+
+    async def toggleRestrictedRole(self, ctx, transaction, requestedRole,
+                                   category):
+        """
+        Assigs a single role to a user with no checks from a category of roles
+        """
+        member = ctx.message.author
+        roles = self.roles.get(category).split(", ")
```
Comment (davidlougheed) [inline cogs/roles.py:61]: same thing with the get()

File: cogs/roles.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        Self-assign a pronoun role to a user. 
+        """
+        await Roles.toggleRole(self, ctx, transaction, pronoun, "pronouns")
+
+    @commands.command()
+    async def field(self, ctx, transaction, field):
+        """
+        Self-assign a field of study role to a user. 
+        """
+        await Roles.toggleRole(self, ctx, transaction, field, "fields")
```
Comment (davidlougheed) [inline cogs/roles.py:95]: not a static fn, call with `self.toggleRole` and skip passing the self arg by hand
same goes for all of these calls

File: config/config.ini (hunk 8 above)
Comment (krubenok) [reply config/config.ini:84]: Yep, goal is to get some in here so that we can test as the default config. The actual config will differ a lot.

File: cogs/roles.py (hunk 3 above)
Comment (krubenok) [reply cogs/roles.py:51]: Going to go with a more generic error message catch all here.

File: cogs/roles.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        # normalize user input to title case
+        requested_role = requested_role.title()
+        role = discord.utils.get(ctx.guild.roles, name=requested_role)
+
+        if (requested_role not in roles):
+            await ctx.send("Invalid Role")
+            return
+
+        if (transaction == "add"):
+            # Allow only a single faculty role per user.
```
Comment (jidicula) [inline cogs/roles.py:74]: This comment could be made a bit more general since this function isn't going to be called just for faculty role assignment.

File: cogs/roles.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            return
+
+        if (transaction == "add"):
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove"):
+            await member.remove_roles(role, reason="Self Requested")
+        else:
+            await ctx.send("Must `add` or `remove` a role")
+
+    async def toggle_unique_role(self, ctx, transaction, requested_role,
```
Comment (jidicula) [inline cogs/roles.py:57]: I'm not sure that "unique" is the right word for naming here – perhaps some use of "precluding" in the name would be more apt? I feel like there might be some set theory or discrete math term that describes this property exactly, but I'm not knowledgeable enough to know what term would be best.

File: cogs/roles.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                old_role = discord.utils.get(ctx.guild.roles,
+                                             name=category_roles)
+                if old_role is not None:
+                    await member.remove_roles(old_role,
+                                              reason="Self Requested")
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove"):
+            await member.remove_roles(role, reason="Self Requested")
+        else:
+            await ctx.send("Must `add` or `remove` a role")
```
Comment (jidicula) [inline cogs/roles.py:85]: I see a lot of duplicated code between these 2 functions – is it possible to factor out the common parts, and functionalize the "precluding" operation in `toggle_unique_role()`?

Another option would be to write a general function for role addition and add an enclosing category around the faculty and year categories to denote them as single-assignment, then include a check in the function to see if the category argument is in the "single-assignment" category.

File: cogs/roles.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Canary. If not, see <https://www.gnu.org/licenses/>.
+
+# discord-py requirements
+import discord
+from discord import utils
+from discord.ext import commands
+
+import codecs
+import configparser
```
Comment (jidicula) [inline cogs/roles.py:26]: Almost forgot – these 2 imports are unused in this file, so they can probably be removed.

File: cogs/roles.py (hunk 12 above)
Comment (krubenok) [reply cogs/roles.py:57]: Restricted? idk asking set theory friends lol.

File: cogs/roles.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        Assigns a single role to a user with no checks from a category of roles
+        """
+        member = ctx.message.author
+        roles = self.roles[category].split(", ")
+
+        # normalize user input to title case
+        requested_role = requested_role.title()
+        role = utils.get(ctx.guild.roles, name=requested_role)
+
+        if (requested_role not in roles):
```
Comment (davidlougheed) [inline cogs/roles.py:48]: ok now im getting pedantic but theres a lot of redundant parens in this diff
im fine with merging without changing these and ill do a cleanup later
